File: src/ArduinoIDE/Picoware/src/internal/gui/scrollbar.cpp
#include "../../internal/gui/scrollbar.hpp"

namespace Picoware
{
    ScrollBar::ScrollBar(Draw *draw, Vector position, Vector size, uint16_t outlineColor, uint16_t fillColor)
        : display(draw), position(position), size(size), outlineColor(outlineColor), fillColor(fillColor)
    {
        this->draw();
    }

    ScrollBar::~ScrollBar()
    {
        // Destructor
    }

    void ScrollBar::clear()
    {
        this->display->clear(position, size, fillColor);
    }

    void ScrollBar::draw()
    {
        this->display->drawRect(position, size, outlineColor);
        this->display->fillRect(Vector(position.x + 1, position.y + 1), Vector(size.x - 2, size.y - 2), fillColor);
    }

    void ScrollBar::setAll(Vector newPosition, Vector newSize, uint16_t newOutlineColor, uint16_t newFillColor, bool shouldDraw, bool shouldClear)
    {
        if (shouldClear)
        {
            this->clear();
        }
        this->position = newPosition;
        this->size = newSize;
        this->outlineColor = newOutlineColor;
        this->fillColor = newFillColor;
        if (shouldDraw)
        {
            this->draw();
        }
    }

    void ScrollBar::setPosition(Vector newPosition, bool shouldDraw, bool shouldClear)
    {
        if (shouldClear)
        {
            this->clear();
        }
        this->position = newPosition;
        if (shouldDraw)
        {
            this->draw();
        }
    }

    void ScrollBar::setSize(Vector newSize, bool shouldDraw, bool shouldClear)
    {
        if (shouldClear)
        {
            this->clear();
        }
        this->size = newSize;
        if (shouldDraw)
        {
            this->draw();
        }
    }

    void ScrollBar::setOutlineColor(uint16_t newColor, bool shouldDraw, bool shouldClear)
    {
        if (shouldClear)
        {
            this->clear();
        }
        this->outlineColor = newColor;
        if (shouldDraw)
        {
            this->draw();
        }
    }

    void ScrollBar::setFillColor(uint16_t newColor, bool shouldDraw, bool shouldClear)
    {
        if (shouldClear)
        {
            this->clear();
        }
        this->fillColor = newColor;
        if (shouldDraw)
        {
            this->draw();
        }
    }
}

File: src/CircuitPython/picoware/gui/scrollbar.py
class ScrollBar:
    """A simple scrollbar class for a GUI."""

    def __init__(
        self,
        draw,
        position,  # Vector
        size,  # Vector
        outline_color: int = 0x0000,
        fill_color: int = 0xFFFFFF,
        is_horizontal: bool = False,
    ) -> None:
        from picoware.system.vector import Vector

        self.display = draw
        self.position = position
        self.size = size
        self.outline_color = outline_color
        self.fill_color = fill_color
        self.is_horizontal = is_horizontal

        self.fill_pos = Vector(self.position.x + 1, self.position.y + 1)
        self.fill_size = Vector(self.size.x - 2, self.size.y - 2)

    def __del__(self):
        if self.position:
            del self.position
            self.position = None
        if self.size:
            del self.size
            self.size = None

    def clear(self) -> None:
        """Clear the scrollbar."""
        self.display.clear(self.position, self.size, self.fill_color)

    def draw(self) -> None:
        """Draw the scrollbar."""
        if self.is_horizontal:
            # Draw horizontal scrollbar
            self.fill_pos.x, self.fill_pos.y = self.position.x + 1, self.position.y + 1
            self.fill_size.x, self.fill_size.y = self.size.x - 2, self.size.y - 2
            self.display.rect(self.position, self.size, self.outline_color)
            self.display.fill_rectangle(
                self.fill_pos,
                self.fill_size,
                self.fill_color,
            )
        else:
            # Draw vertical scrollbar
            self.fill_pos.x, self.fill_pos.y = self.position.x + 1, self.position.y + 1
            self.fill_size.x, self.fill_size.y = self.size.x - 2, self.size.y - 2
            self.display.rect(self.position, self.size, self.outline_color)
            self.display.fill_rectangle(
                self.fill_pos,
                self.fill_size,
                self.fill_color,
            )

    def set_all(
        self,
        position,  # Vector
        size,  # Vector
        outline_color: int,
        fill_color: int,
        is_horizontal: bool = False,
        should_draw: bool = True,
        should_clear: bool = True,
    ) -> None:
        """Set the properties of the scrollbar."""
        if should_clear:
            self.clear()
        self.position = position
        self.size = size
        self.outline_color = outline_color
        self.fill_color = fill_color
        self.is_horizontal = is_horizontal
        if should_draw:
            self.draw()
